SPIRE SDS integration documentation (#11057)
istio/istio.io: content/en/docs/ops/integrations/spire/index.md

---
title: SPIRE
description: Describes how to configure Istio to integrate with SPIRE to get cryptographic identities through Envoy's SDS API.

> weight: 40

Author: I've added 1 to the `zipkin` page and updated `SPIRE` page to be 31, is it ok?

aliases:
owner: istio/wg-networking-maintainers
test: no
---

[SPIRE](https://spiffe.io/docs/latest/spire-about/spire-concepts/) is a production-ready implementation of the SPIFFE specification that performs node and workload attestation in order to securely
issue cryptographic identities to workloads running in heterogeneous environments. SPIRE can be configured as a source of cryptographic identities for Istio workloads through an integration with
[Envoy's SDS API](https://www.envoyproxy.io/docs/envoy/latest/configuration/security/secret). Istio can detect the existence of a UNIX Domain Socket that implements the Envoy SDS API on a defined
socket path, allowing Envoy to communicate and fetch identities directly from it.

This integration with SPIRE provides flexible attestation options not available with the default Istio identity management while harnessing Istio's powerful service management.
For example, SPIRE's plugin architecture enables diverse workload attestation options beyond the Kubernetes namespace and service account attestation offered by Istio.
SPIRE's node attestation extends attestation to the physical or virtual hardware on which workloads run.

> 

Reviewer: Let us make sure we add a warning that Istio 1.14 is required, assuming this feature lands in 1.14.

Reviewer 2: I am not sure I agree? This is the docs for Istio 1.14 - it should be assumed everything needs 1.14?

Reviewer: Thinking more about it, i think it would be useful to note both data plane and control plane are required to be 1.14.  (Is this correct?)

Reviewer: Only would be useful to clarify if this only works with NEW istio 1.14 vs upgrading to Istio 1.14 @alexandrealvino

Author: Hi @linsun, only the data plane and `istioctl` are required to be 1.14. I will clarify that on the docs.

Reviewer: That  sounds great, thanks @alexandrealvino !

    granting them access to the SPIRE Agent's UNIX Domain Socket.

1. Use [sidecar injection](/docs/setup/additional-setup/sidecar-injection) to inject the `istio-proxy`
    container into the pods within your mesh. See [custom templates](/docs/setup/additional-setup/sidecar-injection/#custom-templates-experimental)
    for information on how to apply the custom defined `spire` template into `istio-proxy`. This allows for the CSI driver to mount the UDS on the sidecars.

    Check Ingress-gateway pod state:

    {{< text bash >}}
    $ kubectl get pods -n istio-system
    NAME                                    READY   STATUS    RESTARTS   AGE
    istio-ingressgateway-5b45864fd4-lgrxs   0/1     Running   0          17s
    istiod-989f54d9c-sg7sn                  1/1     Running   0          23s
    {{< /text >}}

Data plane containers will only reach the `Ready` if a corresponding registration entry is created for them on the SPIRE Server. Then,
Envoy will be able to fetch cryptographic identities from SPIRE.
See [Register workloads](#register-workloads) to register entries for services in your mesh.

## Register workloads

> This section describes the options available for registering workloads in SPIRE Server.

Reviewer 3: in **a** SPIRE?

### Option 1: Automatic registration using the SPIRE workload registrar

By deploying [SPIRE Kubernetes Workload Registrar](https://github.com/spiffe/spire/tree/main/support/k8s/k8s-workload-registrar)

> along with SPIRE Server, new entries are automatically registered for each new pod that is created.

Reviewer 3: with **a** SPIRE?

    {{< text bash >}}
$ INGRESS_POD=$(kubectl get pod -l istio=ingressgateway -n istio-system -o jsonpath="{.items[0].metadata.name}")
$ INGRESS_POD_UID=$(kubectl get pods -n istio-system $INGRESS_POD -o jsonpath='{.metadata.uid}')
    {{< /text >}}

1. Get the spire-server pod:

    {{< text bash >}}
$ SPIRE_SERVER_POD=$(kubectl get pod -l app=spire-server -n spire -o jsonpath="{.items[0].metadata.name}")
    {{< /text >}}

1. Register an entry for the SPIRE Agent running on the node:

    {{< text bash >}}
$ kubectl exec -n spire $SPIRE_SERVER_POD -- \
/opt/spire/bin/spire-server entry create \
-spiffeID spiffe://example.org/ns/spire/sa/spire-agent \
-selector k8s_psat:cluster:demo-cluster \
-selector k8s_psat:agent_ns:spire \
-selector k8s_psat:agent_sa:spire-agent \
-node -socketPath /run/spire/sockets/server.sock

Entry ID         : d38c88d0-7d7a-4957-933c-361a0a3b039c
SPIFFE ID        : spiffe://example.org/ns/spire/sa/spire-agent
Parent ID        : spiffe://example.org/spire/server
Revision         : 0
TTL              : default
Selector         : k8s_psat:agent_ns:spire
Selector         : k8s_psat:agent_sa:spire-agent
Selector         : k8s_psat:cluster:demo-cluster

>     {{</ text >}}

Reviewer 3: Lint error is caused by this:
```suggestion
    {{< /text >}}
```
You have a similar issue below.

Author: Nice catch, thanks.

Revision         : 0
TTL              : default
Selector         : k8s:ns:istio-system
Selector         : k8s:`pod-uid`:`63c2bbf5-a8b1-4b1f-ad64-f62ad2a69807`
Selector         : k8s:sa:`istio-ingressgateway-service-account`
DNS name         : `istio-ingressgateway`.istio-system.svc
DNS name         : `istio-ingressgateway-c48554dd6-cff5z`
    {{</ text >}}

1. Deploy an example workload:

    {{< text bash >}}
$ istioctl kube-inject --filename @samples/security/spire/sleep-spire.yaml | kubectl apply -f -
    {{< /text >}}

    Note that the workload will need the SPIFFE CSI Driver volume to access the SPIRE Agent socket. To accomplish this,
    you can leverage the `spire` pod annotation template from [Install Istio](#install-istio) section or add the CSI volume to
    the deployment spec of your workload. Both of these alternatives are highlighted on the example snippet below:

    {{< text yaml >}}
apiVersion: apps/v1
kind: Deployment
metadata:
  name: sleep
spec:
  replicas: 1
  selector:
    matchLabels:
      app: sleep
  template:
    metadata:
      labels:
        app: sleep
      # Injects custom sidecar template
      annotations:

>         inject.istio.io/templates: "sidecar,spire"

Reviewer 4: why spire enable or not shall be aware by every pod/deployment through this annotation? Is there a global flag for it ?Should it be enabled in mesh scope?

Author: The annotation applies the custom `spire` template created on the configuration of the `Install Istio` section:
```
      templates:
        spire: |
          spec:
            containers:
            - name: istio-proxy
              volumeMounts:
              - name: workload-socket
                mountPath: /run/secrets/workload-spiffe-uds
                readOnly: true
             volumes:
               - name: workload-socket
                 csi:
                   driver: "csi.spiffe.io"
```
The custom template feature facilitates the injection of the `spiffe-csi-driver` volume into workload pods. 

But one can always do it manually by adding the volume to the workload spec:
```
      volumes:
      - name: tmp
        emptyDir: {}
      # CSI volume
      - name: workload-socket
        csi:
          driver: "csi.spiffe.io"
```

---
title: Istio CA Integration with SPIRE 

> linktitle: How to integrate Istio with SPIRE through Envoy SDS API.

Reviewer 2: ```suggestion
linktitle: SPIRE
```

To match the others in this folder like `cert-manager`

Reviewer 2: I am fine with either option

---
title: SPIRE

> description: Describes how to configure Istio to integrate with SPIRE to get cryptographic identities through Envoy's SDS API.

Reviewer 3: Nit: The other headings are with `How to ...` or `Information on how to ...`. Maybe change this to the former by removing Describes?

aliases:
owner: istio/wg-networking-maintainers
test: no
---

[SPIRE](https://spiffe.io/docs/latest/spire-about/spire-concepts/) is a production-ready implementation of the SPIFFE specification that performs node and workload attestation in order to securely
issue cryptographic identities to workloads running in heterogeneous environments. SPIRE can be configured as a source of cryptographic identities for Istio workloads through an integration with
[Envoy's SDS API](https://www.envoyproxy.io/docs/envoy/latest/configuration/security/secret). Istio can detect the existence of a UNIX Domain Socket that implements the Envoy SDS API on a defined
socket path, allowing Envoy to communicate and fetch identities directly from it.

This integration with SPIRE provides flexible attestation options not available with the default Istio identity management while harnessing Istio's powerful service management.
For example, SPIRE's plugin architecture enables diverse workload attestation options beyond the Kubernetes namespace and service account attestation offered by Istio.
SPIRE's node attestation extends attestation to the physical or virtual hardware on which workloads run.

> For a quick demo of how this SPIRE integration with Istio works, see [Integrating SPIRE as a CA through Envoy's SDS API]({{< github_tree >}}/samples/security/spire).

Reviewer 3: FYI: Need to merge https://github.com/istio/istio/pull/37948 for this link.

{{< warning >}}

> Note that this integration requires version 1.14 for both `istioctl` and data plane.

Reviewer 3: and **the**  data plane ?

{{< text bash >}}
$ kubectl apply -f @samples/security/spire/spire-quickstart.yaml
{{< /text >}}

This will deploy SPIRE into your cluster, along with two additional components: the [SPIFFE CSI Driver](https://github.com/spiffe/spiffe-csi) — used to share the SPIRE Agent's UNIX Domain Socket with the other
pods throughout the node — and the [SPIRE Kubernetes Workload Registrar](https://github.com/spiffe/spire/tree/main/support/k8s/k8s-workload-registrar), a facilitator that performs automatic workload registration
within Kubernetes. See [Install Istio](#install-istio) to configure Istio and integrate with the SPIFFE CSI Driver.

### Option 2: Configure a custom SPIRE installation

See the [SPIRE's Quick start for Kubernetes guide](https://spiffe.io/docs/latest/try/getting-started-k8s/)
to get started deploying SPIRE into your Kubernetes environment. See [SPIRE CA Integration Prerequisites](#spire-ca-integration-prerequisites)
for more information on configuring SPIRE to integrate with Istio deployments.

#### SPIRE CA Integration Prerequisites

To integrate your SPIRE deployment with Istio, configure SPIRE:

> 1. Access [SPIRE Agent reference](https://spiffe.io/docs/latest/deploying/spire_agent/#agent-configuration-file) and

Reviewer 3: Access the

    configure the SPIRE Agent socket path to match the Envoy SDS defined socket path.

    {{< text plain >}}
    socket_path = "/run/secrets/workload-spiffe-uds/socket"
    {{< /text >}}

1. Share the SPIRE Agent socket with the pods within the node by deploying the
    [SPIFFE CSI Driver](https://github.com/spiffe/spiffe-csi).

See [Install Istio](#install-istio) to configure Istio to integrate with the SPIFFE CSI Driver.

{{< warning >}}
Note that you must deploy SPIRE before installing Istio into your environment so that Istio can detect it as a CA.
{{< /warning >}}

## Install Istio

1. [Download Istio release 1.14+](/docs/setup/getting-started/#download).

> 1. After [deploying SPIRE](#install-spire) into your environment, and verifying that all deployments are in Ready state,

Reviewer 3: You use 'Ready' below. Let's do that here as well.

    install Istio with custom patches for the Ingress-gateway as well as for istio-proxy.

    {{< text bash >}}
    $ istioctl install --skip-confirmation -f - <<EOF
    apiVersion: install.istio.io/v1alpha1
    kind: IstioOperator
    metadata:
      namespace: istio-system
    spec:
      profile: default
      meshConfig:
        trustDomain: example.org
      values:
        global:
        # This is used to customize the sidecar template
        sidecarInjectorWebhook:
          templates:
            spire: |
              spec:
                containers:
                - name: istio-proxy
                  volumeMounts:
                  - name: workload-socket

>                     mountPath: /run/secrets/workload-spiffe-uds

Reviewer 4: @alexandrealvino Why it is /run/secrets/xxx instead of /var/run/secrets/xxx? Why remove the /var prefix? I know this /run/secrets/xxxx is aligned with spire-agent configmap configuration, but istio-agent only check /var/run/secrets/workload-spiffe-uds/socket as following shows, or I missed anything?
```go
    WorkloadIdentitySocketPath = "./var/run/secrets/workload-spiffe-uds/socket"
...
	socketExists, err := checkSocket(ctx, security.WorkloadIdentitySocketPath)
	if err != nil {
		return nil, fmt.Errorf("failed to check SDS socket: %v", err)
	}

	if socketExists {
		log.Info("SDS socket found. Istio SDS Server won't be started")
	} else {
		log.Info("SDS socket not found. Starting Istio SDS Server")
		err = a.initSdsServer()
		if err != nil {
			return nil, fmt.Errorf("failed to start SDS server: %v", err)
		}
	}
```

Reviewer 5: `/var/run` is a symlink to `/run`

Reviewer 4: ok, thanks for clarification.

    {{< text bash >}}
    $ kubectl get pods -n istio-system
    NAME                                    READY   STATUS    RESTARTS   AGE
    istio-ingressgateway-5b45864fd4-lgrxs   0/1     Running   0          17s
    istiod-989f54d9c-sg7sn                  1/1     Running   0          23s
    {{< /text >}}

Data plane containers will only reach `Ready` if a corresponding registration entry is created for them on the SPIRE Server. Then,
Envoy will be able to fetch cryptographic identities from SPIRE.
See [Register workloads](#register-workloads) to register entries for services in your mesh.

## Register workloads

This section describes the options available for registering workloads in a SPIRE Server.

### Option 1: Automatic registration using the SPIRE workload registrar

By deploying [SPIRE Kubernetes Workload Registrar](https://github.com/spiffe/spire/tree/main/support/k8s/k8s-workload-registrar)
along with a SPIRE Server, new entries are automatically registered for each new pod that is created.

See [Verifying that identities were created for workloads](#verifying-that-identities-were-created-for-workloads)
to check issued identities.

> Note that `SPIRE workload registrar` is used on [quick start](#option-1:-quick-start) section.

Reviewer 3: used in the quick start

### Option 2: Manual Registration

To improve workload attestation security robustness, SPIRE is able to verify against a group of selector values based on different parameters. Skip these steps if you installed `SPIRE` by following the [quick start](#option-1:-quick-start) since it uses automatic registration.

> 1. To generate an entry for an Ingress Gateway with a set of selectors such as the

Reviewer 3: ~~To~~ Generate

    {{< text bash >}}
    $ kubectl exec -n spire spire-server-0 -- \
    /opt/spire/bin/spire-server entry create \
        -spiffeID spiffe://example.org/ns/default/sa/sleep \
        -parentID spiffe://example.org/ns/spire/sa/spire-agent \
        -selector k8s:ns:default \
        -selector k8s:pod-uid:$SLEEP_POD_UID \
        -dns $SLEEP_POD \
        -socketPath /run/spire/sockets/server.sock
    {{< /text >}}

{{< warning >}}
SPIFFE IDs for workloads must follow the Istio SPIFFE ID pattern: `spiffe://<trust.domain>/ns/<namespace>/sa/<service-account>`
{{< /warning >}}

See the [SPIRE help on Registering workloads](https://spiffe.io/docs/latest/deploying/registering/) to learn how to create new entries for workloads and get them attested using multiple selectors to strengthen attestation criteria.

## Verifying that identities were created for workloads

Use the following command to confirm that identities were created for the workloads:

{{< text bash >}}
$ kubectl exec -i -t $SPIRE_SERVER_POD -n spire -c spire-server -- /bin/sh -c "bin/spire-server entry show -socketPath /run/spire/sockets/server.sock"
Found 3 entries
Entry ID         : c8dfccdc-9762-4762-80d3-5434e5388ae7

> SPIFFE ID        : spiffe://example.org/ns/istio-system/sa/istio-ingressgateway-service-account

Reviewer 4: @alexandrealvino  How about the TLS communication between istiod and istio-proxy? which kind of certification is used here?

Reviewer 5: The TLS communication between istiod and istio-agent uses a certificate generated by istiod itself, as per default behavior. That communication wasn't changed at all.

Reviewer 4: > The TLS communication between istiod and istio-agent uses a certificate generated by istiod itself, as per default behavior. That communication wasn't changed at all.

Thanks, in fact, my question is why not integrate istiod to SPIRE also. It will look more complete. Or will it be the next step or anything block this?

Reviewer 5: It's the next step.

See [Verifying that identities were created for workloads](#verifying-that-identities-were-created-for-workloads)
to check issued identities.

> Note that `SPIRE workload registrar` used on [quick start](#option-1:-quick-start) section.

Reviewer 6: ```suggestion
Note that `SPIRE workload registrar` is used in the [quick start](#option-1:-quick-start) section.
```
